### ISSUE TITLE
How can I extend to add a dynamically positioned timestamp in the bubble?

### ISSUE THREAD

**Issue by zwang** [jessesquires/JSQMessagesViewController#1290]
Thank you for the great library, I am learning a lot from it and definitely can save us many hours.

I want to do something like the screenshot shows: depending on the text length and the width of the bubble, the timestamp position is different, the second case will require recalculating the bubble size. Any suggestion how I can extend the default behavior for this?  Thank you for your help.
![screen shot 2015-10-28 at 8 57 52 am](https://cloud.githubusercontent.com/assets/1171212/10794508/7c2f4fee-7d52-11e5-884f-cb4b09c801ea.png)
![screen shot 2015-10-28 at 8 58 52 am](https://cloud.githubusercontent.com/assets/1171212/10794509/7c324190-7d52-11e5-9f3a-faa791d4aaa5.png)

To decide whether to put the timestamp in a new line or at the end of the last line, I am calculating the last line white space width by calling this method: 

    var lastUsedRect: CGRect? = nil
    var lastRect: CGRect? = nil
    messageTextView.layoutManager.enumerateLineFragmentsForGlyphRange(NSMakeRange(0, messageData.text!().characters.count)) { (rect, usedRect, textContainer, glyphRange, stopped) -> Void in
            lastUsedRect = usedRect
            lastRect = rect
        }

    if let r = lastRect, u = lastUsedRect {
            let emptyWidth = r.width - u.width
            if emptyWidth > label.frame.width {
                 // No need to add extra height to bubble
            } else {
                 // Add extra height to bubble
            } 
    }

So I either need to be able to access messageTextView in the JSQMessagesCollectionViewFlowLayout or pass the extra height of bubble to the calculating size method in calculator.

**Comment by zwang** [comment]
For people who will read this, I figured it out with a fresh mind in the morning. I just override the following method and update the height to the right height (this is just an example, need to calculate to get 30): 

    override func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        let size = super.collectionView(collectionView, layout: collectionViewLayout, sizeForItemAtIndexPath: indexPath)
        return CGSize(width: size.width, height: size.height + 30)
    }

I like this solution because I avoid go to change JSQMessage's internal implementation. I am trying to extend as much as possible. 

If there is any better solution, please let me know. Thank you!

**Comment by jessesquires** [comment]
Hey @zwang ! Thanks for updating with a solution :smile: 

That looks good to me if it works.

However, my intent with `JSQMessagesBubblesSizeCalculator` was to easily accommodate scenarios like this — to give you a way to avoid overriding methods, and instead provide a new object that does the extra work you need to do.

What limitations did you run into with this? 

------------

Would the following work?

1. Subclass `JSQMessagesBubblesSizeCalculator`.
2. Call super in the `messageBubbleSizeForMessageData: ...` method
3. Adjust and return this value (i.e.,  add 30 points)

If so, I think this is cleaner and fits better for working within the library.

**Comment by zwang** [comment]
Hi @jessesquires, 

Thank you for the insights. I will refactoring to use this approach later. It is definitely better and cleaner solution. I tried it earlier but due to some reasons I gave up that approach (explained below).

Just in case somebody is interested. I tried this approach earlier, but I needed to access the textView to figure out the white space available for the timeStamp so that I can decide whether the height get added 30pt or 0pt. I didn't see an easy way pass that in. So I changed to the approach above. 

However, after some experiment with the approach above, I encountered the "chicken, egg, chicken" problem. I need to get the collectionViewCell to get the textView so I can get the size delta in the overridden method above. 
    override func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize

So I called the 
    super.collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell 
to get the collectionViewCell, which triggered the issue mentioned in this ticket https://github.com/jessesquires/JSQMessagesViewController/issues/819. It seems to be a chicken-egg who comes first problem to me. So I had a work around to create a standalone textView with the same setup as the textView in the cell and put in the msg text to get the height delta value.
